phylo/src/substitution_models/protein_models/mod.rs

```diff
@@ -0,0 +1,180 @@
+use std::fmt::Display;
+
+use crate::evolutionary_models::{EvoModelParams, ProteinModelType};
+use crate::substitution_models::{
+    FreqVector, SubstMatrix, SubstModel, SubstModelInfo, SubstitutionLikelihoodCost,
+    SubstitutionModel,
+};
+use crate::Result;
+
+pub(crate) mod protein_model_generics;
+pub(crate) use protein_model_generics::*;
+
+pub(crate) type ProteinSubstArray = [f64; 400];
+pub(crate) type ProteinFrequencyArray = [f64; 20];
+
+pub type ProteinSubstModel = SubstModel<ProteinSubstParams>;
+pub type ProteinSubstModelInfo = SubstModelInfo<ProteinSubstModel>;
+pub type ProteinLikelihoodCost<'a> = SubstitutionLikelihoodCost<'a, ProteinSubstModel>;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum ProteinParameter {
+    Pi,
+    Mu,
+    Lambda,
+}
+#[derive(Clone, Debug, PartialEq)]
+pub struct ProteinSubstParams {
+    pub(crate) model_type: ProteinModelType,
+    pub(crate) pi: FreqVector,
+}
+
+impl EvoModelParams for ProteinSubstParams {
+    type ModelType = ProteinModelType;
+    type Parameter = ProteinParameter;
+    fn new(model_type: &ProteinModelType, _: &[f64]) -> Result<Self>
+    where
+        Self: Sized,
+    {
+        Ok(Self {
+            model_type: *model_type,
+            pi: match model_type {
+                ProteinModelType::WAG => wag_pi(),
+                ProteinModelType::HIVB => hivb_pi(),
+                ProteinModelType::BLOSUM => blosum_pi(),
+                _ => unreachable!(),
+            },
+        })
+    }
+    fn parameter_definition(
+        _model_type: &ProteinModelType,
+    ) -> Vec<(&'static str, Vec<ProteinParameter>)> {
+        todo!()
+    }
+    fn get_value(&self, _param_name: &ProteinParameter) -> f64 {
+        todo!()
+    }
+    fn set_value(&mut self, _param_name: &ProteinParameter, _value: f64) {
+        todo!()
+    }
+    fn get_pi(&self) -> &FreqVector {
+        &self.pi
+    }
+    fn set_pi(&mut self, _pi: FreqVector) {
+        todo!()
+    }
+}
+
+impl Display for ProteinSubstParams {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.pi)
+    }
+}
+
+impl From<ProteinSubstParams> for Vec<f64> {
+    fn from(val: ProteinSubstParams) -> Self {
+        val.pi.as_slice().to_vec()
```

Comment:
super-nit: this feels strange. probably just needs a comment. I don't know from this code what's the definition of FreqVector, it feels like it's a named vec[f64]. so, basically, it feels like this code should just return a clone? or maybe even just drop `FreqVector` and leave underlying type? again, I don't know what excactly, but it feels strange, and probably needs a comment

Reply:
This is so strange only because, in this case, all the parameters are defined in a vector. For anything other than the protein substitution parameters, this would have more parameters that are not stored in a single vector. And the FreqVector is a DVector<f64> from nalgebra by type, so not the same as Vec<f64>, confusingly enough. That said, I realised this is absolutely unnecessary and I can remove all implementations of this trait from my parameters, which I have now done (: